lib/deas/runner.rb

```diff
@@ -22,24 +28,95 @@ def initialize(handler_class, args = nil)
       @request         = args[:request]
       @session         = args[:session]
       @params          = args[:params] || {}
+
+      @status, @headers, @body = nil, Rack::Utils::HeaderHash.new, nil
+    end
+
+    def to_rack
+      [self.status || DEFAULT_STATUS, self.headers.to_hash, self.body || DEFAULT_BODY]
+    end
+
+    def run
+      raise NotImplementedError
+    end
+
+    def status(value = nil)
+      @status = value if !value.nil?
+      @status
     end
 
-    def halt(*args);           raise NotImplementedError; end
-    def redirect(*args);       raise NotImplementedError; end
-    def content_type(*args);   raise NotImplementedError; end
-    def status(*args);         raise NotImplementedError; end
-    def headers(*args);        raise NotImplementedError; end
-    def send_file(*args);      raise NotImplementedError; end
+    def headers(value = nil)
+      @headers.merge!(value) if !value.nil?
+      @headers
+    end
 
-    # the render methods are used by both the deas and test runners
-    # so we implement here
+    def body(value = nil)
+      if !value.nil?
+        unless value.respond_to?(:each)
+          raise ArgumentError, "this body value (class `#{value.class}`), does not "\
+                               "respond to `each`."
+        end
+        @body = value
+      end
+      @body
+    end
+
+    def content_type(extname, params = nil)
+      self.headers['Content-Type'] = get_content_type(extname, params)
+    end
+
+    def halt(*args)
+      self.status(args.shift)  if args.first.instance_of?(::Fixnum)
+      self.headers(args.shift) if args.first.kind_of?(::Hash)
+      self.body(args.shift)    if args.first.respond_to?(:each)
+      throw :halt
+    end
+
+    def redirect(location, *halt_args)
+      self.status(302)
+      self.headers['Location'] = get_absolute_url(location)
+      halt(*halt_args)
+    end
+
+    def send_file(file_path, opts = nil)
+      path_name = Pathname.new(file_path)
+      self.halt(404, Rack::Utils::HeaderHash.new, []) if !path_name.exist?
+
+      env   = self.request.env
+      mtime = path_name.mtime.httpdate.to_s
+      self.halt(304, Rack::Utils::HeaderHash.new, []) if env['HTTP_IF_MODIFIED_SINCE'] == mtime
```

Comment:
@kellyredding - I'm fine with this because I'm guessing its what Sinatra was doing but it seems a little aggressive to always use an empty headers hash. 

I could maybe see this being annoying if you had like a custom header (one of those "X-" headers) set on all requests (via a `before_init` or something). There might be other issues though with keeping the headers so I'm fine with it using an empty headers hash and addressing this issue later if it ever comes up.